src/perform_step/bdf_perform_step.jl

```diff
@@ -863,7 +873,13 @@ function perform_step!(integrator, cache::QNDFCache{max_order}, repeat_step=fals
   update_D!(D, dd, k)
   if integrator.opts.adaptive
     @unpack abstol, reltol, internalnorm = integrator.opts
-    calculate_residuals!(atmp, dd, uprev, u, abstol, reltol, internalnorm, t)
+    if cache.consfailcnt > 1 && mass_matrix !== I
+      # if we get repeated failure and mass_matrix !== I it's likely that
+      # there's a discontinuity on the algebraic equations
+      calculate_residuals!(atmp, mul!(nlsolver.tmp, mass_matrix, dd), uprev, u, abstol, reltol, internalnorm, t)
```

Comment:
makes sense.

Reply:
FWIW @YingboMa this is exactly the case we are testing in #1606